propolis/src/block/in_memory.rs

```diff
@@ -0,0 +1,279 @@
+use std::collections::VecDeque;
+use std::io::{Error, ErrorKind, Result};
+use std::num::NonZeroUsize;
+use std::sync::{Arc, Condvar, Mutex};
+
+use super::DeviceInfo;
+use crate::block;
+use crate::dispatch::{AsyncCtx, DispCtx, Dispatcher, SyncCtx, WakeFn};
+use crate::inventory::Entity;
+use crate::vmm::SubMapping;
+
+use tokio::sync::Semaphore;
+
+pub struct InMemoryBackend {
+    bytes: Arc<Mutex<Vec<u8>>>,
+
+    driver: Mutex<Option<Arc<Driver>>>,
+    worker_count: NonZeroUsize,
+
+    is_ro: bool,
+    block_size: usize,
+    sectors: usize,
+}
+
+impl InMemoryBackend {
+    pub fn create(bytes: Vec<u8>, is_ro: bool) -> Result<Arc<Self>> {
+        let len = bytes.len();
+
+        let this = Self {
+            bytes: Arc::new(Mutex::new(bytes)),
+
+            driver: Mutex::new(None),
+            worker_count: NonZeroUsize::new(1).unwrap(),
+
+            is_ro,
+            block_size: 512,
+            sectors: len / 512,
```

Comment:
It should probably be documented that a 512b sector is demanded here.

Reply:
made configurable in 6030809